docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,217 @@
+The proposal, as outlined below, suggests to implement the necessary changes in the gardener core components and to adapt the [SeedAuthorizer](https://github.com/gardener/gardener/issues/1723) to consider `Bastion` resources that the Gardener API Server serves.
+
+
+### Goals
+- Operators can request and will be granted **one-time**, **personal** , and **time-limited** `ssh` access to cluster nodes:
+  - **one-time**: Credentials are under Gardener control and can be used only for the extent of this request (of course multiple `ssh` sessions are possible, in parallel or repeatedly, but after "the time is up", see time-limited property, they will be revoked and can never be used ever again; this, btw., prohibits use of private credentials that Gardener cannot permanently revoke/rotate)
+  - **personal**: Credentials are bound to the requestor and logins and logs can be properly audited and traced back to the requestor
+  - **time-limited** Credentials are valid only for a specific time span and cannot be used beyond that time span
+- The respective generated **personal `ssh` credentials will be made available to the requestor** and nobody else can use (a.k.a. impersonate) them (not even other operators).
+- Necessary infrastructure resources for `ssh` access (such as VMs, firewall rules, etc.) are automatically created and also terminated after usage, but at the latest after the above mentioned max time span.
+
+### Non-Goals
+- Node-specific access
+- Auditability on operating system level (not only auditing the SSH login, but everything that is done on a node and other respective resources, e.g. by using dedicated operating system users)
+- Reuse of temporarily created necessary infrastructure resources by different users
+
+## Proposal
+
+### Involved Components
+The following is a list of involved components, that either need to be newly introduced or extended if already existing
+- Gardener API Server (`GAPI`)
+  - New `operations.gardener.cloud` API Group
+  - New resource type `Bastion`, see [resource example](#resource-example) below
+  - New Admission Webhooks for `Bastion` resource
+  - `SeedAuthorizer`: The `SeedAuthorizer` and dependency graph needs to be extended to consider the `Bastion` resource https://github.com/gardener/gardener/tree/master/pkg/admissioncontroller/webhooks/auth/seed/graph
+- `gardenlet`
+  - Deploys `Bastion` CRD under the `extensions.gardener.cloud` API Group to the Seed, see [resource example](#resource-example) below
+  - Similar to `BackupBucket`s or `BackupEntry`, the `gardenlet` watches the `Bastion` resource in the garden cluster and creates a seed-local `Bastion` resource, on which the provider specific bastion controller acts upon
+- `gardenctlv2` (or any other client)
+  - Creates `Bastion` resource in the garden cluster
+  - Establishes an SSH connection to a shoot node, using a bastion host as proxy
+  - Heartbeats / keeps alive the `Bastion` resource during SSH connection
```

Comment:
As said out of band, I'd prefer the bastion host to remain up for longer than the `gardenctl` `ssh` session, so that I can use my own `ssh` tooling and also so that I can run a longer (logical) debugging session during my shift without waiting for the bastion again and again. Maybe a deferred termination 1h after the last heartbeat but at the latest 8h or 12h or whatever after creation of the `bastion` resource makes sense? Basically, an operator should only once have to wait for the bastion to come up, but during his shift not a lot more often than that.

Reply:
> As said out of band, I'd prefer the bastion host to remain up for longer than the gardenctl ssh session, so that I can use my own ssh tooling

Very much +1 on that.

Reply:
What would happen if the user runs `gardenctl ssh ...` multiple times with the same flags? Will every invocation create a new bastion? Will we somehow pool them (i.e. gardenctl checks if a bastion from that user with that key already exists)?

Reply:
Also, if users are expected / projected to use their own tooling, will this mean that they typically do a 

```
gardenctl ssh myshoot node1
// yadda yadda yadda

// and in another terminal then
ssh user@bastion
```

This means that we need to clearly print the SSH command that the user should use. And we need to tell them "your first connection, you do with gardenctl, then every other you can do with whatever tool you have."

Maybe a workflow similar to `kubectl port-forward` would make more sense? Having the `gardenctl ssh` command only establish the tunnel and either keep it open or exit with a `fmt.Print("Done. Tunnel is open for 24h.")`.

Reply:
> What would happen if the user runs `gardenctl ssh ...` multiple times with the same flags? Will every invocation create a new bastion? Will we somehow pool them (i.e. gardenctl checks if a bastion from that user with that key already exists)?

Yes I would propose that `gardenctl` first tries to find an already existing bastion for the shoot (and ssh key) and to reuse it, except maybe when it's called by `--force`

Regarding your second comment:
I think it makes sense to have two different `gardenctl` commands. Maybe:
`ssh` command: creates bastion and afterwards establishes ssh connection (similar to https://github.com/gardener/gardenctl/blob/master/pkg/cmd/ssh_alicloud.go#L113)
`bastion` command: as you proposed, keeps bastion open and heartbeats it until it receives a signal to terminate, writes the ssh key for the node to a temporary directoy and prints out the ssh command to use (similar to https://github.com/gardener/gardenctl/blob/master/pkg/cmd/ssh_alicloud.go#L113). On termination it would cleanup the temporary file for the node ssh key

Reply:
> Yes I would propose that gardenctl first tries to find an already existing bastion for the shoot (and ssh key) and to reuse it, except maybe when it's called by --force

And user, right :)

> I think it makes sense to have two different gardenctl commands. Maybe:

👍🏻 

> writes the ssh key for the node to a temporary directoy and prints out the ssh command to use

Why is this needed? The SSH key is the user's personal key, or? No need to write it into a tmp dir?

Reply:
> And user, right :)

yes :)

> Why is this needed? The SSH key is the user's personal key, or? No need to write it into a tmp dir?

I was referring to the nodes SSH private key. `gardenctl` will first obtain the SSH private key of the node(s) and put it into some temporary directory

Example:
```bash
ssh -o ProxyCommand="ssh -i ~/.ssh/id_rsa -W %h:%p gardener@bastion_host" -i /tmp/nodes.key gardener@shoot_node
```

Reply:
OK, I see. I was wondering about the workflow if I just create the bastion and SSH myself. Should we install the node private key directly onto the bastion already so that a two-step SSH process (not using proxy command) would also be possible?

Reply:
We can have an argument (TBD) for the [`bastion` command](https://github.com/gardener/gardener/pull/3802#discussion_r614336968) to `scp` the private key of the node onto the bastion to enable this scenario. I wouldn't do this by default.